index.bs

```diff
@@ -677,14 +692,515 @@ id=] must be the same as the [=window handle=].
+
+  1. Return the result of getting the value for |object| in [=object id map=].
+</div>
+
+<div algorithm>
+To get the <dfn export>object for an object id</dfn> given an |object id|:
+
+  1. For each |object| → |id value| of [=object id map=]:
+
+    1. If |id value| is equal to |object id|, return [=success=] with data
+       |object|
+
+  1. Return [=error=] with [=error code=] [=no such element=]
+
+  Issue: This error code isn't right.
+</div>
+
+[=remote end definition=] and [=local end definition=]
+```
+RemoteObject = {
+  UndefinedObject //
+  NullObject //
+  StringObject //
+  NumberObject //
+  BooleanObject //
+  BigIntObject //
+  SymbolObject //
+  ArrayObject //
+  FunctionObject //
+  ObjectObject //
+  FunctionObject //
+  RegexObject //
+  DateObject //
+  MapObject //
+  SetObject //
+  WeakMapObject //
+  WeakSetObject //
+  IteratorObject //
+  GeneratorObject //
+  ErrorObject //
+  ProxyObject //
+  PromiseObject //
+  TypedArrayObject //
+  ArrayBufferObject //
+  NodeObject //
+  WindowProxyObject //
+}
+
+ObjectId = text;
+
+UndefinedObject = {
+  type: "undefined",
+  value: null,
+}
+
+NullObject = {
+  type: "null",
+  value: null,
+}
+
+StringObject = {
+  type: "string",
+  value: text,
+}
+
+SpecialNumber = "NaN" / "-0" / "+Infinity" / "-Infinity";
+
+NumberObject = {
+  type: "number",
+  value: number / SpecialNumber,
+}
+
+BooleanObject = {
+  type: "boolean",
+  value: bool,
+}
+
+BigIntObject = {
+  type: "bigint",
+  value: text,
+}
+
+SymbolObject = {
+  type: "symbol",
+  objectId: ObjectId,
+}
+
+ArrayObject = {
+  type: "array",
+  objectId: ObjectId,
+  value?: [RemoteObject]
+}
+
+ObjectObject = {
+  type: "object",
+  objectId: ObjectId,
+  value?: {*text: RemoteObject}
+}
+
+FunctionObject = {
+  type: "function",
+  objectId: ObjectId,
+}
+
+RegexObject = {
+  type: "regex",
+  objectId: ObjectId,
+  value: text
+}
+
+DateObject = {
+  type: "date",
+  objectId: ObjectId,
+  value: text
+}
+
+MapObject = {
+  type: "map",
+  objectId: ObjectId,
+}
+
+SetObject = {
+  type: "set",
+  objectId: ObjectId
+}
+
+WeakMapObject = {
+  type: "map",
+  objectId: ObjectId,
+}
+
+WeakSetObject = {
+  type: "set",
+  objectId: ObjectId,
+}
+
+IteratorObject = {
+  type: "iterator",
+  objectId: ObjectId,
+}
+
+GeneratorObject = {
+  type: "iterator",
+  objectId: ObjectId,
+}
+
+ErrorObject = {
+  type: "error",
+  objectId: ObjectId,
+}
+
+ProxyObject = {
+  type: "proxy",
+  objectId: ObjectId,
+}
+
+PromiseObject = {
+  type: "promise",
+  objectId: ObjectId,
+}
+
+TypedArrayObject = {
+  type: "typedarray",
+  objectId: ObjectId,
+}
+
+ArrayBufferObject = {
+  type: "arraybuffer",
+  objectId: ObjectId,
+}
+
+NodeObject = {
+  type: "node",
+  objectId: ObjectId,
+  value?: NodeValue,
+}
+
+NodeValue = {
+  nodeType: uint,
+  nodeName: text,
+  localName: text,
+  nodeValue: text,
+  childNodeCount: uint,
+  children?: [*NodeObject],
```

Comment:
Not all nodes have children.

Is this inspired by existing protocols? Taking a close look at them would be good. We could try to represent all node types individually, or flatten everything into a single "node". I expect that elements at least are special in existing protocols?

Reply:
CDP just has https://chromedevtools.github.io/devtools-protocol/tot/DOM/#type-Node WebDriver 1.0 only supports elements, not other node types as return values. Firefox RDP is a bit of a different design, but afiact it only has a node grip type.